theano/tensor/signal/pool.py

```diff
@@ -2438,3 +2438,185 @@ def c_code(self, node, name, inp, out, sub):
+
+    def perform(self, node, inp, out):
+        image_data, roi = inp
+        top_data, argmax_data, = out
+        num_roi = roi.shape[0]
+        spatial_scale = self.spatial_scale
+        pool_height = self.pooled_h
+        pool_width = self.pooled_w
+
+        batch_size = image_data.shape[0]
+        n_channels = image_data.shape[1]
+        image_width = image_data.shape[3]
+        assert image_data.ndim == 4
+        assert roi.ndim == 2
+        maxval_coordinates = []
+        max_vals = []
+        for b_in in range(batch_size):
+            for i in range(num_roi):
+                x_start = numpy.floor((roi[i, 1] * spatial_scale) + 0.5)
+                y_start = numpy.floor((roi[i, 2] * spatial_scale) + 0.5)
+                x_end = numpy.floor((roi[i, 3] * spatial_scale) + 0.5)
+                y_end = numpy.floor((roi[i, 4] * spatial_scale) + 0.5)
+
+                roi_height = max(y_end - y_start + 1, 1)
+                roi_width = max(x_end - x_start + 1, 1)
+                row_length = roi_width / pool_width
+                col_length = roi_height / pool_height
+
+                for cn in range(n_channels):
+                    for jy in range(pool_height):
+                        for ix in range(pool_width):
+                            x1 = int(numpy.floor(x_start + ix * row_length))
+                            x2 = int(numpy.ceil(x1 + row_length))
+                            y1 = int(numpy.floor(y_start + jy * col_length))
+                            y2 = int(numpy.ceil(y1 + col_length))
+                            interest_region = image_data[b_in, cn, y1:y2, x1:x2]
+                            max_vals.append(numpy.max(interest_region))
+                            argmax = numpy.unravel_index(numpy.argmax(interest_region), interest_region.shape)
+                            maxloc = argmax[1] + x1 + (argmax[0] + y1) * image_width
+                            maxval_coordinates.append(maxloc)
+        # Reshaped as (batch_index, num_roi, channels, pool_h * pool_w)
+        max_vals = numpy.reshape(numpy.array(max_vals), (batch_size, num_roi, n_channels, pool_height * pool_width))
+        maxval_coordinates = numpy.reshape(numpy.array(maxval_coordinates), (batch_size, num_roi, n_channels, pool_height * pool_width))
+        top_data[0] = max_vals
+        argmax_data[0] = maxval_coordinates
+
+    def infer_shape(self, node, in_shapes):
+        data_shape = tensor.shape(node.inputs[0])
+        rois_shape = tensor.shape(node.inputs[1])
+        batch_size = data_shape[0]
+        num_rois = rois_shape[0]
+        h = self.pooled_h
+        w = self.pooled_w
+        channels = data_shape[1]
+        out_shape = [batch_size, num_rois, channels, h * w]
+        return [out_shape, out_shape]
+
+    def grad(self, inp, grads):
+        data, roi = inp
+        gz1, gz2 = grads
+        maxout, argmax = self(data, roi)
```

Comment:
Now we have a way to not request this duplicate node. Can you try to use that new feature? For this, change

    def grad(....

to

    def L_op(self, inputs, outputs, gout):

and you can do:

    maxout, argmax = outputs 

instead of rebuilding it.

Reply:
Makes sense and nice feature indeed! I'll change it